### PR TITLE
Fix issue 19893 - extern(C++ 'ns') should count as module scope

### PR DESCRIPTION
**Pull request by Geod24**
Alternative implementation to #9891 
@thewilsonator : That's why I was asking were `addMember` was called from.
CC @TurkeyMan : That might break some of your code (See the fix to cppmangle3).

### CODE REVIEW

**Comment by thewilsonator** [comment]
Does this work for nested namespaces? e.g.
```
extern(C++, "foo")
extern(C++, "bar")
version = baz;
```

**Comment by Geod24** [comment]
Yes

**Comment by Geod24** [comment]
Also, there's probably some wrong lookup code we can remove after this, but I'll take a look at completely deprecating non-mangle Nspace.

File: test/compilable/cppmangle3.d [hunk 1]
```diff
@@ -34,5 +34,22 @@ struct Foo
 alias Alias(alias a) = a;
 alias Alias(T) = T;
 
-static assert(is(Alias!(__traits(parent, bar)) == Foo));
```
Comment (thewilsonator) [inline test/compilable/cppmangle3.d]: How did this originally pass? 
also `static assert(is(Alias!(__traits(parent, Foo.bar)) == Foo));` currently fails
as does `Foo.bar` (with `Error: no property bar for type Foo`). You should update the bugzilla issue or create a new one for this.

**Review by thewilsonator — APPROVED** [review]
mod bugzilla issues.

File: test/compilable/cppmangle3.d (hunk 1 above)
Comment (Geod24) [reply test/compilable/cppmangle3.d]: https://github.com/dlang/dmd/pull/8721

Comment (thewilsonator) [reply test/compilable/cppmangle3.d]: I mean How can you even refer to `bar` without going through `Foo`?

Comment (Geod24) [reply test/compilable/cppmangle3.d]: The symbol resolution code using `toParent3` was a bit to eager I guess.
I tried reverting that PR code fully, but it fails on this test case.
I'll try to deprecate non-mangle `extern(C++, scoped)` soon. There's a tons of bugs in the implementation (e.g. it shows up in `__traits(allMembers)` which prevents me from automatically checking if my bindings are correct...).

Comment (thewilsonator) [reply test/compilable/cppmangle3.d]: Awesome, looking forward to it.

**Comment by TurkeyMan** [comment]
What should I expect this to break?

**Comment by thewilsonator** [comment]
https://issues.dlang.org/show_bug.cgi?id=19920

Please update title and commit title to close that bug and target stable.

**Comment by Geod24** [comment]
@thewilsonator : Thanks. Updated.

**Comment by thewilsonator** [comment]
Please target stable that parent bug is pretty severe.

**Comment by Geod24** [comment]
But it might also break code (and has been in 2.084)

**Comment by thewilsonator** [comment]
What code do you think it might break? I think that code would be broken anyway. 

Also ou typo'd your commit message so the bot isn't picking up on it:
`Fix issue_e_ 19920 - Wrong parent for `extern(C++, "ns")` nested in scope`

**Comment by Geod24** [comment]
Fixed the commit title.

> What code do you think it might break?

The very same code that is changed in the first commit: Using `bar` instead of `Foo.bar`.

> I think that code would be broken anyway.

Absolutely not.
The notion of "broken code" varies from individual to individual, but here's mine: If it never does what any individual expects it to do, it is broken. In this case, it's very unlikely to be done "by mistake".

**Comment by thewilsonator** [comment]
I would consider that code broken, and I would consider it broken by that definition also. In this case thing I would except nobody to expect is for it to compile: that code ought to give a "no such identifier `bar`" error.

**Comment by Geod24** [comment]
> In this case thing I would except nobody to expect is for it to compile

But it *did* compile and we're breaking that. It's not great but this feature is not widely used and it's a very niche usage. But let's not do it in a point release...

**Comment by Geod24** [comment]
This broke:
```d
extern(C++, "namespace")
{
    struct Foo {}
    void func(Foo);
}
pragma(msg, func.mangleof);
```

This used to be mangled as   `_ZN9namespace4funcENS_3FooE`
After this PR it is mangled as `_ZN9namespace4funcE3Foo`
Note the lack of namespace on `Foo`.
Scary this is not covered by the test suite. The scoped (`extern(C++, namespace)`) is covered, and I discovered this while trying to deprecate it and switch over tests.
Will fix soon.

**Comment by TurkeyMan** [comment]
I'm really surprised the string namespace tests didn't catch this O_o

**Pull request merged**